File: src/ARC/ARCtrl.fs
namespace ARCtrl

open FileSystem
open Contract
open ISA

open Fable.Core

[<AttachMembers>]
type ARC =
    {
       ISA : ISA.Investigation
       CWL : CWL.CWL
       FileSystem : FileSystem.FileSystem 
    }    

    static member create(
        isa: ISA.Investigation,
        cwl: CWL.CWL,
        fs: FileSystem.FileSystem 
    ) =
        {
           ISA = isa
           CWL = cwl
           FileSystem = fs
        }  

    //static member updateISA (isa : ISA.Investigation) (arc : ARC) : ARC =
    //    raise (System.NotImplementedException())

    //static member updateCWL (cwl : CWL.CWL) (arc : ARC) : ARC =
    //    raise (System.NotImplementedException())

    //static member updateFileSystem (fileSystem : FileSystem.FileSystem) (arc : ARC) : ARC =
    //    raise (System.NotImplementedException())

    //static member addFile (path : string) (arc : ARC) : ARC =
    //    FileSystem.addFile |> ignore
    //    ARC.updateFileSystem |> ignore
    //    raise (System.NotImplementedException())

    //static member addFolder (path : string) (arc : ARC) : ARC =
    //    FileSystem.addFolder |> ignore
    //    ARC.updateFileSystem |> ignore
    //    raise (System.NotImplementedException())

    //static member addFolders (paths : string array) (arc : ARC) : ARC =
    //    paths
    //    |> Array.fold (fun arc path -> ARC.addFolder path arc) arc |> ignore
    //    raise (System.NotImplementedException())

    ///// Add folder to ARC.FileSystem and add .gitkeep file to the folder
    //static member addEmptyFolder (path : string) (arc : ARC) : ARC =
    //    FileSystem.addFolder  |> ignore   
    //    FileSystem.addFile (Path.combine path Path.gitKeepFileName) |> ignore
    //    ARC.updateFileSystem |> ignore
    //    raise (System.NotImplementedException())

    //static member addEmptyFolders (paths : string array) (arc : ARC) : ARC =
    //    paths
    //    |> Array.fold (fun arc path -> ARC.addEmptyFolder path arc) arc |> ignore
    //    raise (System.NotImplementedException())


    /// Add assay folder to the ARC.FileSystem and update the ARC.ISA with the new assay metadata
    //static member addAssay (assay : ISA.ArcAssay) (studyIdentifier : string) (arc : ARC) : ARC = 
  
        // - Contracts - //
        // create spreadsheet assays/AssayName/isa.assay.xlsx  
        // create text assays/AssayName/dataset/.gitkeep 
        // create text assays/AssayName/dataset/Readme.md
        // create text assays/AssayName/protocols/.gitkeep 
        // update spreadsheet isa.investigation.xlsx

        // - ISA - //
        //let assayFolderPath = Path.combineAssayFolderPath assay
        //let assaySubFolderPaths = Path.combineAssaySubfolderPaths assay
        //let assayReadmeFilePath = Path.combine assayFolderPath Path.assayReadmeFileName
        //let updatedInvestigation = ArcInvestigation.addAssay assay studyIdentifier arc.ISA
        //arc
        // - FileSystem - //
        //// Create assay root folder in ARC.FileSystem
        //ARC.addFolder assayFolderPath arc 
        //// Create assay subfolders in ARC.FileSystem
        //|> ARC.addEmptyFolders assaySubFolderPaths 
        // Create assay readme file in ARC.FileSystem
        //ARC.addFile assayReadmeFilePath 
        //// Update ARC.ISA with the updated investigation
        //|> ARC.updateISA updatedInvestigation 

    // to-do: we need a function that generates only create contracts from a ARC data model. 
    // reason: contracts are initially designed to sync disk with in-memory model while working on the arc.
    // but we need a way to create an arc programmatically and then write it to disk.

    // to-do: function that returns read contracts based on a list of paths.
    // the list of paths is used to create a filesystem tree
    static member createReadContracts (filePaths : string array) : Contract array =
        let xlsxReadContractFromPath (path: string) = Contract.createRead(path, DTOType.Spreadsheet)
        let fs = FileSystem.fromFilePaths filePaths
        let xlsxFileNames = [|Path.AssayFileName; Path.StudyFileName; Path.InvestigationFileName|]
        let xlsxFiles = fs.Tree.Filter(fun p -> xlsxFileNames |> Array.contains p)
        match xlsxFiles with
        | Some xlsxPaths -> xlsxPaths.ToFilePaths() |> Array.map xlsxReadContractFromPath
        | None -> [||]

    [<NamedParams(fromIndex=1)>]
    static member createFromReadContracts (cArr: Contract [], ?enableLogging: bool) =
        let enableLogging = defaultArg enableLogging false
        /// filter to only keep *fullfilled* READ contracts of type spreadsheet.
        let filteredContracts = cArr |> Array.choose (fun c ->
            match c with
            | {Operation = READ; DTOType = Some DTOType.Spreadsheet; DTO = Some (DTO.Spreadsheet fsworkbook); Path = p} ->
                Some (p, fsworkbook)
            | _ -> None
        )
        /// get investigation from xlsx
        let investigation = 
            filteredContracts 
            |> Array.find (fun c -> fst c |> FileSystem.Path.isFile Path.InvestigationFileName)
            |> snd 
            |> ISA.Spreadsheet.ArcInvestigation.fromFsWorkbook
        /// get studies from xlsx
        let studies =
            filteredContracts 
            |> Array.filter (fun c -> fst c |> FileSystem.Path.isFile Path.StudyFileName)
            |> Array.map (snd >> ISA.Spreadsheet.ArcStudy.fromFsWorkbook)
        /// get assays from xlsx
        let assays =
            filteredContracts 
            |> Array.filter (fun c -> fst c |> FileSystem.Path.isFile Path.AssayFileName)
            |> Array.map (snd >> ISA.Spreadsheet.ArcAssay.fromFsWorkbook)
        /// Create a investigation copy to check for registered studies/assays
        let copy = investigation.Copy()
        /// Get all registered studies
        let registeredStudies = copy.StudyIdentifiers
        registeredStudies |> Seq.iter (fun studyRegisteredIdent ->
            /// Try find registered study in parsed READ contracts
            let studyOpt = studies |> Array.tryFind (fun s -> s.Identifier = studyRegisteredIdent)
            match studyOpt with
            | Some study -> // This study element is parsed from FsWorkbook and has no regsitered assays, yet
                if enableLogging then printfn "Found study: %s" studyRegisteredIdent
                let registeredAssays = copy.GetStudy(studyRegisteredIdent).AssayIdentifiers
                registeredAssays |> Seq.iter (fun assayRegisteredIdent ->
                    /// Try find registered assay in parsed READ contracts
                    let assayOpt = assays |> Array.tryFind (fun a -> a.Identifier = assayRegisteredIdent)
                    match assayOpt with
                    | Some assay -> 
                        if enableLogging then printfn "Found assay: %s - %s" studyRegisteredIdent assayRegisteredIdent 
                        study.AddAssay(assay)
                    | None -> 
                        if enableLogging then printfn "Unable to find registered assay '%s' in fullfilled READ contracts!" assayRegisteredIdent
                )
                investigation.SetStudy(studyRegisteredIdent, study)
            | None -> 
                if enableLogging then printfn "Unable to find registered study '%s' in fullfilled READ contracts!" studyRegisteredIdent
        )
        investigation

File: ts/ARCtrl.ts
import { tryFind, exactlyOne, choose, map } from "./fable_modules/fable-library-ts/Array.js";
import { fromFsWorkbook } from "./ISA/ISA.Spreadsheet/ArcAssay.js";
import { FsWorkbook } from "./fable_modules/FsSpreadsheet.3.1.1/FsWorkbook.fs.js";
import { ArcAssay } from "./ISA/ISA/ArcTypes/ArcAssay.js";
import { tryFromContract } from "./Contracts/Contracts.ArcAssay.js";
import { Contract } from "./Contract/Contract.js";
import { defaultArg, value, Option } from "./fable_modules/fable-library-ts/Option.js";
import { fromFsWorkbook as fromFsWorkbook_1 } from "./ISA/ISA.Spreadsheet/ArcStudy.js";
import { ArcStudy } from "./ISA/ISA/ArcTypes/ArcStudy.js";
import { tryFromContract as tryFromContract_1 } from "./Contracts/Contracts.ArcStudy.js";
import { fromFsWorkbook as fromFsWorkbook_2 } from "./ISA/ISA.Spreadsheet/InvestigationFile/Investigation.js";
import { tryFromContract as tryFromContract_2 } from "./Contracts/Contracts.ArcInvestigation.js";
import { ArcInvestigation_$reflection, ArcInvestigation } from "./ISA/ISA/ArcTypes/ArcInvestigation.js";
import { FileSystem_$reflection, FileSystem as FileSystem_1 } from "./FileSystem/FileSystem.js";
import { tryISAReadContractFromPath } from "./Contracts/Contracts.ARCtrl.js";
import { iterate } from "./fable_modules/fable-library-ts/Seq.js";
import { printf, toConsole } from "./fable_modules/fable-library-ts/String.js";
import { Record } from "./fable_modules/fable-library-ts/Types.js";
import { IEquatable } from "./fable_modules/fable-library-ts/Util.js";
import { record_type, unit_type, option_type, TypeInfo } from "./fable_modules/fable-library-ts/Reflection.js";

export function ARCAux_getArcAssaysFromContracts(contracts: Contract[]): ArcAssay[] {
    return map<FsWorkbook, ArcAssay>(fromFsWorkbook, choose<Contract, FsWorkbook>(tryFromContract, contracts));
}

export function ARCAux_getArcStudiesFromContracts(contracts: Contract[]): ArcStudy[] {
    return map<FsWorkbook, ArcStudy>(fromFsWorkbook_1, choose<Contract, FsWorkbook>(tryFromContract_1, contracts));
}

export function ARCAux_getArcInvestigationFromContracts(contracts: Contract[]): ArcInvestigation {
    return fromFsWorkbook_2(exactlyOne<FsWorkbook>(choose<Contract, FsWorkbook>(tryFromContract_2, contracts)));
}

export class ARC extends Record implements IEquatable<ARC> {
    readonly ISA: Option<ArcInvestigation>;
    readonly CWL: Option<void>;
    readonly FileSystem: Option<FileSystem_1>;
    constructor(ISA: Option<ArcInvestigation>, CWL: Option<void>, FileSystem: Option<FileSystem_1>) {
        super();
        this.ISA = ISA;
        this.CWL = CWL;
        this.FileSystem = FileSystem;
    }
    static create(isa?: ArcInvestigation, cwl?: Option<void>, fs?: FileSystem_1): ARC {
        return new ARC(isa, cwl, fs);
    }
    static fromFilePaths(filePaths: string[]): ARC {
        const fs: FileSystem_1 = FileSystem_1.fromFilePaths(filePaths);
        return ARC.create(void 0, void 0, fs);
    }
    addFSFromFilePaths(filePaths: string[]): ARC {
        const this$: ARC = this;
        return new ARC(this$.ISA, this$.CWL, FileSystem_1.fromFilePaths(filePaths));
    }
    getReadContracts(): Contract[] {
        const this$: ARC = this;
        const matchValue: Option<FileSystem_1> = this$.FileSystem;
        if (matchValue == null) {
            throw new Error("Cannot create READ contracts from ARC without FileSystem.\r\n\r\nYou could initialized your ARC with `ARC.fromFilePaths` or run `yourArc.addFSFromFilePaths` to avoid this issue.");
        }
        else {
            const fs: FileSystem_1 = value(matchValue);
            return choose<string, Contract>(tryISAReadContractFromPath, fs.Tree.ToFilePaths());
        }
    }
    addISAFromContracts(contracts: Contract[], enableLogging?: boolean): ARC {
        const this$: ARC = this;
        const enableLogging_1: boolean = defaultArg<boolean>(enableLogging, false);
        const investigation: ArcInvestigation = ARCAux_getArcInvestigationFromContracts(contracts);
        const studies: ArcStudy[] = ARCAux_getArcStudiesFromContracts(contracts);
        const assays: ArcAssay[] = ARCAux_getArcAssaysFromContracts(contracts);
        const copy: ArcInvestigation = investigation.Copy();
        iterate<string>((studyRegisteredIdent: string): void => {
            const studyOpt: Option<ArcStudy> = tryFind<ArcStudy>((s: ArcStudy): boolean => (s.Identifier === studyRegisteredIdent), studies);
            if (studyOpt == null) {
                if (enableLogging_1) {
                    toConsole(printf("Unable to find registered study \'%s\' in fullfilled READ contracts!"))(studyRegisteredIdent);
                }
            }
            else {
                const study: ArcStudy = value(studyOpt);
                if (enableLogging_1) {
                    toConsole(printf("Found study: %s"))(studyRegisteredIdent);
                }
                iterate<string>((assayRegisteredIdent: string): void => {
                    const assayOpt: Option<ArcAssay> = tryFind<ArcAssay>((a: ArcAssay): boolean => (a.Identifier === assayRegisteredIdent), assays);
                    if (assayOpt == null) {
                        if (enableLogging_1) {
                            toConsole(printf("Unable to find registered assay \'%s\' in fullfilled READ contracts!"))(assayRegisteredIdent);
                        }
                    }
                    else {
                        const assay: ArcAssay = value(assayOpt);
                        if (enableLogging_1) {
                            toConsole(printf("Found assay: %s - %s"))(studyRegisteredIdent)(assayRegisteredIdent);
                        }
                        study.AddAssay(assay);
                    }
                }, copy.GetStudy(studyRegisteredIdent).AssayIdentifiers);
                investigation.SetStudy(studyRegisteredIdent, study);
            }
        }, copy.StudyIdentifiers);
        return new ARC(investigation, this$.CWL, this$.FileSystem);
    }
}

export function ARC_$reflection(): TypeInfo {
    return record_type("ARCtrl.ARC", [], ARC, () => [["ISA", option_type(ArcInvestigation_$reflection())], ["CWL", option_type(unit_type)], ["FileSystem", option_type(FileSystem_$reflection())]]);
}

